chapters/classes.tex

```diff
@@ -276,36 +276,111 @@ \subsubsection{Prefix Rules}\label{prefix-rules}
   function.
 \end{itemize}
 
-\subsection{Acyclic Bindings of Constants and Parameters}\label{acyclic-bindings-of-constants-and-parameters}
+\subsection{Acyclic Bindings of Constants and Structural Parameters}\label{acyclic-bindings-of-constants-and-parameters}
 
-The unexpanded binding equations for parameters and constants in the translated model must be acyclic after flattening; except that cycles are allowed if the cycles disappear when evaluating parameters having annotation \lstinline!Evaluate = true! that are not part of the cycle.
-Thus it is not possible to introduce equations for parameters by cyclic dependencies.
+The unexpanded binding equations for constants in the translated model must be acyclic after flattening.
+Thus it is not possible to introduce equations for constants by cyclic dependencies.
 
-\begin{nonnormative}
-There is no exception for parameters with \lstinline!fixed = false!, despite the fact that such parameters are generally allowed to be initialized from systems of dependent equations.
-% Saying "binding equation" below even though it's called a "declaration equation", for consistency with the normative paragraph above.
-However, a parameter with \lstinline!fixed = false! can use an initial equation instead of a binding equation, allowing for cyclic dependencies.
-\end{nonnormative}
+Structural parameters, on the other hand, cannot have cyclic dependencies by definition (see \cref{component-variability-prefixes-discrete-parameter-constant}).
```

Comment:
The change in this section is not good at all. The current rule means that you can just look at parameter-bindings and see if there are any cycles or not - which is easy to understand, implement and explain. Restricting it to structural parameters makes it more complicated - and in particular I cannot see how it handles the example below if p and q were parameters.

Reply:
It's handled like this:
```
parameter Real p = 2 * q; /* Normal parameter, due to cycle with q. */
parameter Real q = sin(p); /* Normal parameter, due to cycle with p. */
```

The point is that we don't need acyclic binding for normal parameters, as their value isn't determined until initialization time.

I was going to add a non-normative text explaining that it's a good idea to associate a reason with every parameter being normal (as indicated by the comments in the listing above).  Then, consider a user making a variability error:
```
parameter Integer n = integer(p); /* Normal parameter, declaration equation has parameter variability. */
Real[n] x; /* Error: dimension is not structural expression. */
```
Now, the error message for the non-structural expression can include the reasons why first `n`, and then `p` are normal parameters.

Reply:
I have now added a non-normative text on generation of good error messages.

Reply:
Right, I should have made a comment when pushing [the fix](https://github.com/modelica/ModelicaSpecification/pull/2754/commits/8ba996136b0f3742e152b1faec57e3dc2f24739b).

@eshmoylova, would you agree to marking this as resolved?

Reply:
Yes. Resolving.